wasm/src/main/java/com/dylibso/chicory/wasm/io/BufferWasmInputStream.java

```diff
@@ -0,0 +1,88 @@
+package com.dylibso.chicory.wasm.io;
+
+import java.nio.BufferUnderflowException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+/**
+ * A WASM input stream backed by a byte buffer.
+ */
+class BufferWasmInputStream extends WasmInputStream {
+    private final ByteBuffer buffer;
+
+    BufferWasmInputStream(final ByteBuffer buffer) {
+        super();
+        // make our position independent, and zero-based
+        this.buffer = buffer.slice();
+        this.buffer.order(ByteOrder.LITTLE_ENDIAN);
+    }
+
+    public long position() {
+        return buffer.position();
+    }
+
+    public void skip(final long count) {
+        if (count <= 0) {
+            return;
+        }
+        long newOffs = buffer.position() + count;
+        if (newOffs > buffer.limit()) {
+            throw new WasmEOFException();
+        }
+        buffer.position((int) newOffs);
+    }
+
+    public int rawByteOpt() {
+        return buffer.hasRemaining() ? Byte.toUnsignedInt(buffer.get()) : -1;
```

Comment:
Do we need the `hasRemaining`? Is it breaking on any test?

Reply:
Of course. Please read the documentation of the base class method. If the end-of-file is reached, then -1 shall be returned, else the next byte in the stream is returned. Without `hasRemaining`, an exception will be thrown at EOF, which is contrary to the specification of the method that this is implementing.

Reply:
Solid explanation, works for me, thanks for elaborating 👍